### ISSUE TITLE
[MSL] Support sparse image residency operations in MSL 2.2

### ISSUE THREAD

**Issue by cdavis5e** [KhronosGroup/SPIRV-Cross#1014]
MSL 2.2 on iOS at least adds new methods for reading and sampling textures that additionally return whether or not the texture memory was resident. These correspond to `OpImageSparseXxx` in SPIR-V. It should now be possible to implement those operations.

**Comment by cdavis5e** [comment]
It looks like Apple removed this incomplete feature from the MSL 2.2 headers. Maybe the next version will support it.

**Comment by oscarbg** [comment]
yep.. but we should reopen as it should have returned today in Xcode 11 GM as it's a new feature of A13 GPU..
see (https://twitter.com/gavkar/status/1171494278545018882)

**Comment by cdavis5e** [comment]
They are indeed restored to the MSL 2.2 headers. I guess Apple didn't want anybody poking at this while they finished up this super-secret feature.

**Comment by Degerz** [comment]
Any format limitations with sparse textures on A13 ?

**Comment by cdavis5e** [comment]
> Any format limitations with sparse textures on A13 ?

There might be. See `-[MTLDevice sparseTileSizeWithTextureType:pixelFormat:sampleCount]`.

> Also, it seems to be [implied](https://twitter.com/gavkar/status/1171494806419148800) that the tile mappings are updated from the GPU rather than the CPU ? Is that true ?

That's true. It's actually true of Vulkan as well--see `vkQueueBindSparse()`.

**Comment by Degerz** [comment]
> There might be. See -[MTLDevice sparseTileSizeWithTextureType:pixelFormat:sampleCount:].

I wish Apple would just publicly document the allowed pixel formats like Microsoft does with D3D12 instead. It'd be way more useful this way to see if Metal supports sparse volume textures or sparse MSAA textures ... 

I see that there's an individual `sampleCount` parameter for [sparseTileSize](https://developer.apple.com/documentation/metal/mtldevice/3377856-sparsetilesize) so there might be hope after all to support MSAA textures on at least Intel devices but it would be a big disappointment if all configurations returned 1 ... 

> That's true. It's actually true of Vulkan as well--see vkQueueBindSparse().

I don't mean being able to update the page table into device memory since that's really [ugly](https://gpuopen.com/performance-tweets-series-resource-creation/) performance wise. I want the ability to specifically update tile mappings [via](https://forum.beyond3d.com/posts/1868571/) from the GPU but it looks like that'll remain exclusive to Intel HW so I guess I misinterpreted the tweet ...

**Comment by oscarbg** [comment]
any update on this?

**Comment by Wangsh09** [comment]
any update on this?
